src/details/ArborX_DetailsDistributedTreeImpl.hpp

```diff
@@ -671,6 +678,141 @@ DistributedTreeImpl<DeviceType>::queryDispatch(
+      });
+
+  Distributor<DeviceType> distributor(comm);
+  auto const n_imports = distributor.createFromSends(space, ranks);
+
+  Kokkos::View<QueriesWithIndices<Query> *, typename DeviceType::memory_space>
+      imported_queries_with_indices(
+          Kokkos::view_alloc(
+              space, Kokkos::WithoutInitializing,
+              "ArborX::DistributedTree::query::imported_queries_with_indices"),
+          n_imports);
+
+  sendAcrossNetwork(space, distributor, exported_queries_with_indices,
+                    imported_queries_with_indices);
+
+  // Execute the callback on the process owning the primitives.
+  OutputView remote_out(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::DistributedTree::query::remote_out"),
+      n_imports);
+  KokkosExt::reallocWithoutInitializing(space, indices, n_imports);
+  Kokkos::parallel_for(
+      "ArborX::DistributedTree::query::execute_callbacks",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0,
+                                          imported_queries_with_indices.size()),
+      KOKKOS_LAMBDA(int i) {
+        callback(imported_queries_with_indices(i).query,
+                 imported_queries_with_indices(i).primitive_index,
+                 [&](typename OutputView::value_type const &value) {
+                   remote_out(i) = value;
+                   indices(i) = imported_queries_with_indices(i).query_id;
+                 });
+      });
+
+  // Send the result back to the process owning the predicates.
+  Distributor<DeviceType> back_distributor(comm);
+  auto const &dest = distributor.get_sources();
+  auto const &off = distributor.get_source_offsets();
+
+  Kokkos::View<int const *, Kokkos::HostSpace> host_destinations(dest.data(),
+                                                                 dest.size());
+  Kokkos::View<int const *, Kokkos::HostSpace> host_offsets(off.data(),
+                                                            off.size());
+  typename DeviceType::memory_space memory_space;
+  Kokkos::View<int *, DeviceType> destinations(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::DistributedTree::query::destinations"),
+      dest.size());
+  Kokkos::deep_copy(space, destinations, host_destinations);
+  Kokkos::View<int *, DeviceType> offsets(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::DistributedTree::query::offsets"),
+      off.size());
+  Kokkos::deep_copy(space, offsets, host_offsets);
+  auto const n_imports_back =
+      back_distributor.createFromSends(space, destinations, offsets);
+  KokkosExt::reallocWithoutInitializing(space, out, n_imports_back);
+  Kokkos::View<int *, DeviceType> query_ids(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::DistributedTree::query::nearest::query_ids"),
+      n_imports_back);
+
```

Comment:
This could be simplified by using `UnmanagedView` from the distributor data pointers, and doing an immediate deep_copoy.

Reply:
Isn't that what I'm doing already?